Abilities API: Add the initial core abilities and categories registration.

jorgefilipecosta opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/64146


Backports the initial WordPress core abilities proposed at https://github.com/WordPress/abilities-api/pull/108 with improvements in preparation for core at https://github.com/WordPress/abilities-api/pull/128, into Core WordPress.

## Testing
Verify the unit tests are passing.
Loaded the block editor, oppened the browser console, exectuted the following two calls:
```
wp.apiFetch({ path: wp.url.addQueryArgs('/wp-abilities/v1/abilities/core/get-site-info/run' ) } ).then(console.log);
wp.apiFetch({ path: wp.url.addQueryArgs('/wp-abilities/v1/abilities/core/get-site-info/run', {input: {fields: ['name','url' ] } } ) } ).then(console.log);
```
Verified things returned as expected.
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jorgefilipecosta committed:
Add: Core abiltiies registration.

jorgefilipecosta committed:
correct version field

jorgefilipecosta committed:
lint fix

gziolo commented on tests/phpunit/tests/abilities-api/wpCoreAbilities.php:
```diff
@@ -0,0 +1,135 @@
+		$this->assertArrayHasKey( 'wp_version', $ability_data );
+		$this->assertSame( $environment, $ability_data['environment'] );
+
+		wp_set_current_user( 0 );
```
This might be not needed because `tear_down` from `WP_UnitTestCase_Base` does that.

jorgefilipecosta replied:
Removed 👍

jorgefilipecosta committed:
fix wpRegisterAbility tests

gziolo commented on tests/phpunit/tests/abilities-api/wpCoreAbilities.php:
```diff
@@ -0,0 +1,135 @@
+
+		$ability = wp_get_ability( 'core/get-user-info' );
+
+		$this->assertTrue( $ability->check_permissions() );
```
I don't think it matter here, but `execute` calls the same check internally, so this test checks it twice. However, it's fine to assert it expliclitly.

jorgefilipecosta replied:
Yes, I guess by asserting it explicitly we guard against future changes.

gziolo commented on tests/phpunit/tests/abilities-api/wpCoreAbilities.php:
```diff
@@ -0,0 +1,135 @@
+		$this->assertIsArray( $result );
+		$this->assertArrayHasKey( 'name', $result );
+		$this->assertArrayHasKey( 'url', $result );
+		$this->assertArrayNotHasKey( 'description', $result );
```
Nit: maybe checking the count of fields would be simpler instead of looking for all missing keys.

jorgefilipecosta replied:
Changed 👍

gziolo commented on src/wp-includes/abilities/wp-core-abilities.php:
```diff
@@ -0,0 +1,271 @@
+					'destructive' => false,
+					'idempotent'  => true,
+				),
+				'show_in_rest' => false,
```
I'm not sure I fully understand why this is hidden from REST API, as this could be useful for agentic loops on the client. However, there was probably some decision process that I didn't notice.

jorgefilipecosta replied:
I also don't have the context of why it has hidden, but wanted to kept is his, as showing is possible in the future while hiding after showing may not be possible.

jorgefilipecosta committed:
additional test fixes

gziolo commented on src/wp-includes/abilities/wp-core-abilities.php:
```diff
@@ -0,0 +1,271 @@
+
+declare( strict_types = 1 );
+/**
+ * Registers the core abilities categories.
```
Nit: there was a larger discussion around the naming and the plural form should look like:

```suggestion
 * Registers the core ability categories.
```

gziolo reviewed (approved):
Nice one, thank you for working on syncing this code from Abilities API repository. I see you applied the feedback from @aaronjorbin regarding how the code is structured 👍🏻

jorgefilipecosta committed:
in progress test fixes

jorgefilipecosta committed:
in progress test fixes

mukeshpanchal27 commented on tests/phpunit/tests/abilities-api/wpCoreAbilities.php:
```diff
@@ -0,0 +1,135 @@
+<?php declare( strict_types=1 );
```
```suggestion
<?php

declare( strict_types=1 );
```

mukeshpanchal27 commented on tests/phpunit/tests/abilities-api/wpCoreAbilities.php:
```diff
@@ -0,0 +1,135 @@
+	/**
+	 * Tests that the `core/get-site-info` ability is registered with the expected schema.
+	 */
+	public function test_core_get_bloginfo_ability_is_registered(): void {
```
Please add `@ticket 64146` to the test that was added as part of this PR.

jorgefilipecosta replied:
Done 👍

mukeshpanchal27 commented on tests/phpunit/tests/abilities-api/wpRegisterAbilityCategory.php:
```diff
@@ -52,6 +62,10 @@ public function tear_down(): void {
 	public function test_register_category_before_init_hook(): void {
+		// simulate that the 'wp_abilities_api_categories_init' action has not yet fired.
+		global $wp_actions;
+		$wp_actions[ 'wp_abilities_api_categories_init' ] = 0;
```
```suggestion
		$wp_actions['wp_abilities_api_categories_init'] = 0;
```

Fix PHPCS

jorgefilipecosta replied:
Done 👍

gziolo commented on tests/phpunit/tests/abilities-api/wpAbilitiesRegistry.php:
```diff
@@ -34,6 +34,9 @@ public function set_up(): void {
 		}
 
+		// Remove core registration action to prevent re-registration.
+		remove_action( 'wp_abilities_api_categories_init', 'wp_register_core_ability_categories' );
```
It might make more sense to remove this default actions from all tests in the usual place. We do similar tricks for example for core blocks.

jorgefilipecosta replied:
Done 👍

jorgefilipecosta committed:
middle most tests working

jorgefilipecosta committed:
one failure

jorgefilipecosta committed:
fix remaning test case

jorgefilipecosta committed:
Update tests/phpunit/tests/abilities-api/wpCoreAbilities.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

jorgefilipecosta committed:
Update src/wp-includes/abilities/wp-core-abilities.php

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

jorgefilipecosta committed:
test fixes different appraoch

jorgefilipecosta committed:
Fix abilities API tests without test-specific changes

- Modified tests/phpunit/includes/functions.php to unhook core abilities at priority 1 (before registration) instead of priority 1000 (after)
- Updated tests/phpunit/tests/abilities-api/wpCoreAbilities.php to use set_up_before_class() for one-time core abilities registration
- All 239 abilities API tests now pass (565 assertions)
- Minimal test-specific code required

jorgefilipecosta committed:
add ticket mark to new tests

jorgefilipecosta committed:
apply test feedback

jorgefilipecosta committed:
remove duplicate variable

jorgefilipecosta committed:
remove overprotective unnecessary check

jorgefilipecosta committed:
Add since tag missed on set_up_before_class

jorgefilipecosta committed:
make test names match ability names they are testing
